File: foolish-core/src/main/java/org/foolish/fvm/ubc/BraneMemory.java
package org.foolish.fvm.ubc;

import org.foolish.fvm.BraneMemoryInterface;

import java.util.regex.Pattern;

public class BraneMemory implements BraneMemoryInterface<BraneMemory.Query, FIR> {
    public interface Query{}
    final public class IdentifierQuery extends CharacterizedIdentifier implements Query {
        public IdentifierQuery(String name, String characterization) {
            super(name, characterization);
        }
    }
    final public class RegExpQuery implements Query {
        private final Pattern pattern;
        public RegExpQuery(String regex) {
            this.pattern = Pattern.compile(regex);
        }
    }

    BraneMemory parent;
    private final java.util.Map<Integer, java.util.Map<CharacterizedIdentifier, FIR>> memory;

    public BraneMemory(BraneMemory parent) {
        this.parent=parent;
        this.memory = new java.util.HashMap<>();
    }

    @Override
    public FIR get(Query id, int fromLine) {
        for (int line = fromLine; line >= 0; line--) {
            java.util.Map<CharacterizedIdentifier, FIR> lineMemory = memory.get(line);
            if (lineMemory != null && lineMemory.containsKey(id)) {
                return lineMemory.get(id);
            }
        }
        return null; // Not found
    }

    @Override
    public void put(Query id, FIR value, int byLine) {
        if (id instanceof CharacterizedIdentifier charId) {
            memory.computeIfAbsent(byLine, k -> new java.util.HashMap<>()).put(charId, value);
        } else {
            throw new IllegalArgumentException("Only IdentifierQuery is supported for put operations");
        }
    }}

File: foolish-core/src/main/scala/org/foolish/fvm/scubc/BraneMemory.scala
package org.foolish.fvm.scubc

import org.foolish.fvm.BraneMemoryInterface
import scala.collection.mutable
import java.util.regex.Pattern

/**
 * BraneMemory implementation for the UBC system.
 */
class BraneMemory(val parent: BraneMemory = null) extends BraneMemoryInterface[BraneMemory.Query, FIR]:

  private val memory = mutable.Map[Int, mutable.Map[CharacterizedIdentifier, FIR]]()

  def get(id: BraneMemory.Query, fromLine: Int): FIR =
    (fromLine to 0 by -1).view.flatMap { line =>
      memory.get(line).flatMap { lineMemory =>
        val charId = id.asInstanceOf[CharacterizedIdentifier]
        if lineMemory.contains(charId) then Some(lineMemory(charId))
        else None
      }
    }.headOption.orNull

  def put(id: BraneMemory.Query, value: FIR, byLine: Int): Unit =
    id match
      case charId: CharacterizedIdentifier =>
        memory.getOrElseUpdate(byLine, mutable.Map()).put(charId, value)
      case _ =>
        throw IllegalArgumentException("Only IdentifierQuery is supported for put operations")

object BraneMemory:
  trait Query

  /** Regular class to avoid case-to-case inheritance prohibition */
  class IdentifierQuery(name: String, characterization: String = "")
    extends CharacterizedIdentifier(name, characterization) with Query

  case class RegExpQuery(regex: String) extends Query:
    val pattern: Pattern = Pattern.compile(regex)
